kmod/src/block.c

```diff
@@ -193,117 +202,184 @@ static void block_free(struct super_block *sb, struct block_private *bp)
 	else
 		BUG();
 
-	/* lru_entry could have been clobbered by union member free_node */
 	WARN_ON_ONCE(!list_empty(&bp->dirty_entry));
 	WARN_ON_ONCE(atomic_read(&bp->refcount));
 	WARN_ON_ONCE(atomic_read(&bp->io_count));
 	kfree(bp);
 }
 
 /*
- * We free blocks in task context so we can free kernel virtual mappings.
+ * Free all the blocks that were put in the free_llist.  We have to wait
+ * for rcu grace periods to expire to ensure that no more rcu hash list
+ * lookups can see the blocks.
  */
 static void block_free_work(struct work_struct *work)
 {
-	struct block_info *binf = container_of(work, struct block_info,
-					       free_work);
+	struct block_info *binf = container_of(work, struct block_info, free_work);
 	struct super_block *sb = binf->sb;
 	struct block_private *bp;
 	struct llist_node *deleted;
 
+	scoutfs_inc_counter(sb, block_cache_free_work);
+
 	deleted = llist_del_all(&binf->free_llist);
+	synchronize_rcu();
 
 	llist_for_each_entry(bp, deleted, free_node) {
 		block_free(sb, bp);
 	}
 }
 
 /*
- * After we've dropped the final ref kick off the final free in task
- * context.  This happens in the relatively rare cases of IO errors,
- * stale cached data, memory pressure, and unmount.
+ * Get a reference to a block while holding an existing reference.
+ */
+static void block_get(struct block_private *bp)
+{
+	WARN_ON_ONCE((atomic_read(&bp->refcount) & ~BLOCK_REF_INSERTED) <= 0);
+
+	atomic_inc(&bp->refcount);
+}
+
+/*
+ * Get a reference to a block as long as it's in the hash table.
+ */ 
+static struct block_private *block_get_hashed(struct block_private *bp)
```

Comment:
I added the _if_, yeah, that's clearer.  I now see the hashed/inserted disconnect too, so I changed it to get_if_inserted.  Basing it on the actions, instead of the structure, means we get the nice if (get_if_inserted()) { remove() } blocks which looks pretty clean to me.